### ISSUE TITLE
Feature request: gRPC Support

### ISSUE THREAD

**Issue by 4F2E4A2E** [foss42/apidash#14]
Thank you for this great repo. I would love to have gRPC support.
Is this on your roadmap?

**Comment by ashitaprasad** [comment]
Hi @4F2E4A2E,
Thank you for raising this issue. We definitely have gRPC support on our roadmap.

**Comment by khanjasir90** [comment]
@ashitaprasad @animator would like to work on this issue, is there anything prior we have in place for gRPC or do we have to start it from scratch?

**Comment by ashitaprasad** [comment]
@khanjasir90 There is no prior work on this issue. It involves starting from scratch and building the feature end to end right from the UI design stage to architecture and implementation.

**Comment by khanjasir90** [comment]
@ashitaprasad got it! so will start from doing an analysis of how its done any other tools i.e postman etc and creating mock ups for the same and will share a doc with you.

**Comment by animator** [comment]
Sounds good @khanjasir90 👍

**Comment by SubstantialCattle5** [comment]
hey if possible can i also work with @khanjasir90 on this issue?? 
@animator @ashitaprasad

**Comment by animator** [comment]
@SubstantialCattle5 we do not assign issues. You can discuss/collaborate or work on your own on the issue and post design mockups and discuss how you will implement it in this thread.

**Comment by khanjasir90** [comment]
Hey @animator @ashitaprasad https://docs.google.com/document/d/1R3JVD3Vpp7DhkfkIg6PMwwlYhkXfDaus/edit?usp=drivesdk&ouid=105972227945564036728&rtpof=true&sd=true
Shared above is the details of implementation of grpc feature in API dash.
Please go through it once and provide ur take if any improvement needed.
Thanks 🙏.

**Comment by Ayush0Chaudhary** [comment]
hey I have made a proposal in this Idea, how to get my proposal reviewed? @animator @ashitaprasad

**Comment by WrathOP** [comment]
Hi @ashitaprasad can I work on this gRPC implementation or there's some discussion you guys want to have internally before that.

**Comment by ashitaprasad** [comment]
@WrathOP you can look into https://github.com/foss42/apidash/issues/116 first before taking up gRPC

**Comment by MohamedKamal000** [comment]
If I were to implement it, should I also implement gRPC streaming? or just focus on main features first like choosing services and sending requests and maybe add streaming later

**Comment by MohamedKamal000** [comment]
i have made my plan on how i would implement this feature and i'd appreciate your feedback  @animator  @ashitaprasad 
https://drive.google.com/file/d/1CTvKYiANPd79P2p7G1hAL58WjScr9SAM/view?usp=sharing
feel free to ask anything

**Comment by shreeyansh64** [comment]
I'd like to help with the gRPC implementation. @MohamedKamal000 your approach looks solid, let me know if you want to split the work. @ashitaprasad @animator

**Comment by MohamedKamal000** [comment]
@shreeyansh64  Yeah, sure, I don't mind working together on this, but wait for the owner's review first.

**Comment by animator** [comment]
> i have made my plan on how i would implement this feature and i'd appreciate your feedback [@animator](https://github.com/animator) [@ashitaprasad](https://github.com/ashitaprasad) https://drive.google.com/file/d/1CTvKYiANPd79P2p7G1hAL58WjScr9SAM/view?usp=sharing feel free to ask anything

@MohamedKamal000 Please add your plan directly in the markdown format in this issue thread. We do not review external docs.

**Comment by MohamedKamal000** [comment]
# gRPC feature support for API-Dash

## feature key points and how I would implement them:

- ### Parsing proto files:  
  I will use google `descriptor.proto` file to generate what is called `FileDescriptor` dart class  
  which is a way of parsing a proto file that will also help us in the input mapping process  
- ### Reflection:  
  I will use the google `reflection.proto` file to generate a reflection class that enables us to do the following:  
  + list available services  
  + search for a specific file by name  
  + search for a file which contains a specific services `fileContainingSymbol`  
  Note: the generated reflection class relies on the grpc dart package
- ### Importing files:  
  I will rely on protoc compiler which is basically a .exe file with different versions for different os that we can  
  run based on the running os with specific arguments to generate a `FileDescriptorSet` that contains the  
  FileDescriptor of each proto file  
- ### Sending Requests:
  will parse the user json input and map it to the method message definition he tries to invoke  
  and then encode the input in the same way specified in the protobuff documentation (you can find it at  
  references part) will also decode the response in the same way mentioned in the protobuff documentation  
  and map it to the method message return type definition.  
- ### Streaming:
  the google descriptor file has 2 fields that can help us at the parsing step to decide if a method has client  
  streaming or server streaming or both set to true and the grpc package itself can do streaming through  
  createStreamingCall method  
## good use case scenarios:
1 - writing the server with the ":" at the end to specify the port number,  
    if not specified use default which is 443 and 80 if TLS is not enabled.  
    (also has the option to choose to use TLS or not)  
2 - user have the option to either check for services via reflection or by importing .proto file  
3 - the user choose the method he want to invoke, (unary call, client streaming, server streaming, bidirectional
streaming) among all the methods available for a specific service  
4 - the user will write the request in the form of json (we can have an option that shows the user an example of the
request)  
5 - based on the type of the call the user choose, we will show the request sent and the response in a different way  
    for unary call, we will show the request and response once  
    for client streaming, we will show the different requests sent and the response once  
    for server streaming, we will show the request once and the different responses received  
6- for bidirectional streaming, we will show the different requests sent and the different responses received

## bad use case scenarios and how to handle them (this is just general ones):  
1 - user specify a wrong server address or port number  
sol: log the error in the terminal window saying the server is unreachable  
2 - server has no reflection service enabled and user choose to check via reflection  
sol: log the error in the terminal window saying reflection service is not enabled on the server  
3 - user import a .proto file that is not compatible with the server  
sol: send the request and show the error received from the server ( which in this case will be 12 UNIMPLEMENTED )  
4 - user writes a bad request form (missing required fields, wrong field types, extra fields)  
sol: extra fields will be discarded while showing a warning in the terminal window. missing required fields and
wrong field types will force the user to rewrite the request before sending it.  

### Ui Mockup:
<img width="1151" height="698" alt="Image" src="https://github.com/user-attachments/assets/66b202ea-1d54-4f03-963d-317ed3bf2cef" />

### Implementation plan:

##### I have built good part of the backend side of the feature covering these points:  
###### what is made so far:  
1 - protobuff encoding  
2 - organizing the FileDescriptor data in a manner that can help us in the input mapping step  
3 - reflection and file importing  
4 - sending requests  
5 - serializing json input to a proto message  
6 - deserializing message to json (work for message definition only but will need to refactor it once we implement decoding)  

###### what need to be done:  
1 - proto buff decoding  
2 - raising errors for certain cases (will probably integrate it also with the app terminal window once we integrate
everything)  
3 - dealing with proto dependencies (I already made it but need more testing)  
4 - handle metadata sending  
5 - support streaming  

My plan will be to continue to build the backend side and test it enough, Ui of the app then integrate everything.  

### References:  
the file I used to generate dart code for FileDescriptors, it helped me a lot to understand different proto types  
https://github.com/protocolbuffers/protobuf/blob/main/src/google/protobuf/descriptor.proto  

This is the proto file I used to generate the dart reflection code  
https://github.com/grpc/grpc-proto/blob/master/grpc/reflection/v1/reflection.proto  

proto buff encoding (the site contains useful different sections)  
https://protobuf.dev/programming-guides/encoding/

**Comment by animator** [comment]
you can proceed with the PR @MohamedKamal000

**Comment by PrajwalBenedictNorman** [comment]
Hey @animator 
I’m planning to contribute a small, mergeable slice towards gRPC support to avoid a large, hard-to-review PR.

I’ll start with .proto file import and service/method discovery, and include basic unary gRPC request execution (no streaming yet), with simple error handling and tests. This should provide a clean foundation for expanding gRPC support incrementally (streaming, metadata, auth) in follow-ups.

If this scope works for you, I’ll send a PR soon and follow up with incremental improvements.

**Comment by roshan-melvin** [comment]
Hi @animator sir and @ashitaprasad mam,

As part of addressing the feedback for a visual demonstration in [#1529 ](https://github.com/foss42/apidash/pull/1613), here is the video showing the native gRPC protocol support implemented in API Dash for this issue:

https://github.com/user-attachments/assets/681055a7-a043-4c53-bb59-ceea35bd8b48

You can also check out the new PR ([#1613](https://github.com/foss42/apidash/pull/1613)), which bundles this implementation and includes all the requested video demos in the description. Please let me know if you would like me to cover any specific gRPC methods or scenarios!

Thanks!
